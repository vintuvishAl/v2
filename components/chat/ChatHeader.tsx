import { Ionicons } from "@expo/vector-icons";
import React, { useCallback } from "react";
import { Text, TouchableOpacity, View } from "react-native";
import { useHapticFeedback } from "../../hooks/useHapticFeedback";

interface ChatHeaderProps {
  onMenuPress: () => void;
  onSearchPress: () => void;
  onNewChatPress: () => void;
  useAgent: boolean;
  onToggleAgent: (enabled: boolean) => void;
}

export const ChatHeader: React.FC<ChatHeaderProps> = React.memo(({
  onMenuPress,
  onSearchPress,
  onNewChatPress,
  useAgent,
  onToggleAgent,
}) => {
  const { onButtonPress, onImportantAction } = useHapticFeedback();

  const handleMenuPress = useCallback(() => {
    // Add haptic feedback for menu button
    onButtonPress();
    onMenuPress();
  }, [onButtonPress, onMenuPress]);

  const handleNewChatPress = useCallback(() => {
    // Add haptic feedback for new chat button
    onImportantAction();
    onNewChatPress();
  }, [onImportantAction, onNewChatPress]);

  return (
    <View className="flex-row justify-between items-center px-6 py-4 bg-app-dark-background">
      <View className="flex-row items-center">        <TouchableOpacity testID="all-chats-button" className="p-2 " onPress={handleMenuPress}>
          <Ionicons
            name="menu"
            size={30}
            className="text-app-dark-icon"
            color={"#1cb8ff"}
          />
        </TouchableOpacity>
        <TouchableOpacity
          testID="add-button"
          className="p-2"
          onPress={handleNewChatPress}
        >
          <Ionicons
            name="add"
            size={30}
            className="text-app-dark-icon"
            color={"#1cb8ff"}
          />
        </TouchableOpacity>        {/* <TouchableOpacity testID="search-button" className="p-2" onPress={onSearchPress}>
          <Ionicons
            name="search"
            size={30}
            className="text-app-dark-icon"
            color={"#1cb8ff"}
          />
        </TouchableOpacity> */}
        
      </View>

      <View className="flex-row items-center space-x-2">
        <Text className="text-app-dark-text text-xl font-bold">
          V2 Chat
        </Text>      </View>
    </View>
  );
});

ChatHeader.displayName = 'ChatHeader';
