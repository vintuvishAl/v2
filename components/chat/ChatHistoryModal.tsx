import { Ionicons } from "@expo/vector-icons";
import { FlashList } from "@shopify/flash-list";
import React, { useCallback } from "react";
import { Modal, Text, TouchableOpacity, View } from "react-native";
import Animated, {
  useAnimatedStyle,
  useSharedValue,
  withSpring,
  withTiming
} from 'react-native-reanimated';
import type { Id } from "../../convex/_generated/dataModel";
import { useHapticFeedback } from "../../hooks/useHapticFeedback";

interface ChatHistoryItem {
  _id: Id<"chats">;
  title: string;
  createdAt: number;
  updatedAt?: number;
}

interface ChatHistoryModalProps {
  visible: boolean;
  onClose: () => void;
  chatHistory: ChatHistoryItem[];
  onChatSelect: (chatId: Id<"chats">) => void;
  onDeleteChat: (chatId: Id<"chats">) => void;
  onNewChat: () => void;
  loading?: boolean;
}

export const ChatHistoryModal: React.FC<ChatHistoryModalProps> = ({
  visible,
  onClose,
  chatHistory,
  onChatSelect,
  onDeleteChat,
  onNewChat,
  loading = false,
}) => {
  const { onButtonPress, onImportantAction, onCriticalAction } = useHapticFeedback();
  
  // Optimized date formatting with memoization
  const formatDate = useCallback((timestamp: number) => {
    const date = new Date(timestamp);
    const now = new Date();
    const diffInHours = (now.getTime() - date.getTime()) / (1000 * 60 * 60);
    
    if (diffInHours < 24) {
      return date.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
    } else if (diffInHours < 24 * 7) {
      return date.toLocaleDateString([], { weekday: 'short', hour: '2-digit', minute: '2-digit' });
    } else {
      return date.toLocaleDateString([], { month: 'short', day: 'numeric' });
    }
  }, []);

  // Animated values for smooth modal transitions
  const translateY = useSharedValue(1000);
  const opacity = useSharedValue(0);

  // Update animation when modal visibility changes
  React.useEffect(() => {
    if (visible) {
      translateY.value = withSpring(0, { damping: 20, stiffness: 300 });
      opacity.value = withTiming(1, { duration: 200 });
    } else {
      translateY.value = withTiming(1000, { duration: 200 });
      opacity.value = withTiming(0, { duration: 200 });
    }
  }, [visible, translateY, opacity]);

  // Animated styles
  const backgroundStyle = useAnimatedStyle(() => ({
    opacity: opacity.value,
  }));

  const modalStyle = useAnimatedStyle(() => ({
    transform: [{ translateY: translateY.value }],
  }));

  const handleChatPress = useCallback((chatId: Id<"chats">) => {
    onButtonPress();
    onChatSelect(chatId);
    onClose();
  }, [onButtonPress, onChatSelect, onClose]);

  const handleNewChat = useCallback(() => {
    onImportantAction();
    onNewChat();
    onClose();
  }, [onImportantAction, onNewChat, onClose]);

  const handleDeleteChat = useCallback((chatId: Id<"chats">) => {
    onCriticalAction();
    onDeleteChat(chatId);
  }, [onCriticalAction, onDeleteChat]);

  const handleClose = useCallback(() => {
    onButtonPress();
    onClose();
  }, [onButtonPress, onClose]);

  // Memoized render function for chat items to prevent unnecessary re-renders
  const renderChatItem = useCallback(({ item }: { item: ChatHistoryItem }) => (
    <View className="flex-row items-center px-4 py-3 border-b border-gray-700">
      <TouchableOpacity
        className="flex-1 mr-3"
        onPress={() => handleChatPress(item._id)}
      >
        <Text className="text-white text-base font-medium mb-1" numberOfLines={1}>
          {item.title}
        </Text>
        <Text className="text-gray-400 text-sm">
          {formatDate(item.updatedAt || item.createdAt)}
        </Text>
      </TouchableOpacity>
      <TouchableOpacity
        className="p-2"
        onPress={() => handleDeleteChat(item._id)}
      >
        <Ionicons name="trash-outline" size={18} color="#ef4444" />
      </TouchableOpacity>
    </View>
  ), [handleChatPress, handleDeleteChat, formatDate]);

  // Memoized key extractor
  const keyExtractor = useCallback((item: ChatHistoryItem) => item._id, []);

  return (
    <Modal
      visible={visible}
      transparent={true}
      animationType="none"
      onRequestClose={onClose}
    >
      <Animated.View style={[{ flex: 1, backgroundColor: 'rgba(0,0,0,0.5)' }, backgroundStyle]}>
        <Animated.View style={[{ flex: 1, backgroundColor: '#111827' }, modalStyle]}>
          <View className="flex-row justify-between items-center px-4 py-4 border-b border-gray-700">
            <Text className="text-white text-xl font-semibold">Chat History</Text>
            <TouchableOpacity onPress={handleClose}>
              <Ionicons name="close" size={24} color="white" />
            </TouchableOpacity>
          </View>
          <TouchableOpacity
            className="flex-row items-center px-4 py-4 bg-blue-600 mx-4 mt-4 rounded-lg"
            onPress={handleNewChat}
          >
            <Ionicons name="add" size={20} color="white" />
            <Text className="text-white text-base font-medium ml-2">Start New Chat</Text>
          </TouchableOpacity>
          <View className="flex-1 mt-4">
            {loading ? (
              <View className="flex-1 justify-center items-center">
                <Text className="text-gray-400">Loading chat history...</Text>
              </View>
            ) : chatHistory.length === 0 ? (
              <View className="flex-1 justify-center items-center px-4">
                <Ionicons name="chatbubbles-outline" size={64} color="#6b7280" />
                <Text className="text-gray-400 text-lg mt-4 mb-2">No chat history</Text>
                <Text className="text-gray-500 text-center">
                  Start a conversation to see your chat history here
                </Text>
              </View>
            ) : (              <FlashList
                data={chatHistory}
                renderItem={renderChatItem}
                keyExtractor={keyExtractor}
                showsVerticalScrollIndicator={false}
                estimatedItemSize={70}
                removeClippedSubviews={true}
                getItemType={() => 'chat-item'}
              />
            )}
          </View>
        </Animated.View>
      </Animated.View>
    </Modal>
  );
};
