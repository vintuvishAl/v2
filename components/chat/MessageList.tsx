import React from "react";
import { Text, View } from "react-native";
import Markdown from "react-native-markdown-display";
import type { Message } from "../../hooks/useChat";

interface MessageListProps {
  messages: Message[];
}

export const MessageList: React.FC<MessageListProps> = ({ messages }) => {  // Deduplicate messages based on content and user type to prevent double rendering
  const deduplicatedMessages = React.useMemo(() => {
    const result: Message[] = [];
    const aiMessages: Message[] = [];
    
    // Separate user and AI messages
    messages.forEach(message => {
      if (message.isUser) {
        result.push(message);
      } else {
        aiMessages.push(message);
      }
    });
    
    // For AI messages, remove duplicates and partial duplicates
    const processedAIMessages: Message[] = [];
    aiMessages.forEach(message => {
      // Check if this message is a subset of any existing message
      const isSubsetOfExisting = processedAIMessages.some(existing => 
        existing.text.includes(message.text) && existing.text.length > message.text.length
      );
      
      if (!isSubsetOfExisting) {
        // Remove any existing messages that are subsets of this message
        const filteredExisting = processedAIMessages.filter(existing => 
          !(message.text.includes(existing.text) && message.text.length > existing.text.length)
        );
        
        // Check for exact duplicates in the filtered list
        const isDuplicate = filteredExisting.some(existing => 
          existing.text.trim() === message.text.trim()
        );
        
        if (!isDuplicate) {
          processedAIMessages.length = 0; // Clear array
          processedAIMessages.push(...filteredExisting, message);
        }
      }
    });
    
    // Combine user messages with deduplicated AI messages, sorted by timestamp
    const combined = [...result, ...processedAIMessages];
    return combined.sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());
  }, [messages]);
  const markdownStyles = {
    body: {
      color: '#f3f4f6', // text-app-dark-text color
      fontSize: 16,
    },
    code_inline: {
      backgroundColor: '#374151',
      color: '#e5e7eb',
      paddingHorizontal: 4,
      paddingVertical: 2,
      borderRadius: 4,
    },
    code_block: {
      backgroundColor: '#374151',
      color: '#e5e7eb',
      padding: 12,
      borderRadius: 8,
      marginVertical: 8,
    },
    blockquote: {
      backgroundColor: '#1f2937',
      borderLeftColor: '#6b7280',
      borderLeftWidth: 4,
      paddingLeft: 12,
      paddingVertical: 8,
      marginVertical: 8,
    },
    heading1: {
      color: '#f3f4f6',
      fontSize: 24,
      fontWeight: 'bold' as const,
      marginVertical: 8,
    },
    heading2: {
      color: '#f3f4f6',
      fontSize: 20,
      fontWeight: 'bold' as const,
      marginVertical: 6,
    },
    heading3: {
      color: '#f3f4f6',
      fontSize: 18,
      fontWeight: 'bold' as const,
      marginVertical: 4,
    },
    list_item: {
      color: '#f3f4f6',
      marginVertical: 2,
    },
    strong: {
      color: '#f3f4f6',
      fontWeight: 'bold' as const,
    },
    em: {
      color: '#f3f4f6',
      fontStyle: 'italic' as const,
    },
  };
  return (
    <View className="py-4">
      {deduplicatedMessages.map((message) => (        <View
          key={message.id}
          className={`mb-4 ${
            message.isUser ? "items-end" : "items-start"
          }`}
          testID={`message-${message.id}`}
        >
          <View
            className={`px-4 py-2 rounded-lg ${
              message.isUser
                ? "bg-app-dark-user-message ml-auto max-w-xs lg:max-w-md"
                : "bg-app-dark-ai-message mr-auto w-full"
            }`}
          >
            {message.isUser ? (
              <Text className="text-app-dark-text">{message.text}</Text>
            ) : (
              <Markdown style={markdownStyles}>
                {message.text}
              </Markdown>
            )}
          </View>
        </View>
      ))}
    </View>
  );
};
