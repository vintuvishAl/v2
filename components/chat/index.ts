export { ChatHeader } from "./ChatHeader";
export { ChatHistoryModal } from "./ChatHistoryModal";
export { ChatInput } from "./ChatInput";
export { MessageList } from "./MessageList";
export { ModelSelectorModal } from "./ModelSelectorModal";
export { WelcomeScreen } from "./WelcomeScreen";

