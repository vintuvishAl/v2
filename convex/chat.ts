import { PersistentTextStreaming } from "@convex-dev/persistent-text-streaming";
import { v } from "convex/values";
import { components, internal } from "./_generated/api";
import type { Id } from "./_generated/dataModel";
import { action, internalAction, internalMutation, internalQuery, mutation, query } from "./_generated/server";
import { utilityAgent } from "./agents";

const persistentTextStreaming = new PersistentTextStreaming(components.persistentTextStreaming);

export const createChat = mutation({
  args: {
    prompt: v.string(),
    model: v.optional(v.string()),
  },
  handler: async (ctx, args) => {
    const streamId = await persistentTextStreaming.createStream(ctx);

    const chatId = await ctx.db.insert("chats", {
      title: args.prompt.slice(0, 50) + (args.prompt.length > 50 ? "..." : ""),
      prompt: args.prompt,
      streamId,
      createdAt: Date.now(),
    });

    // Trigger streaming via internal action
    const selectedModel = args.model || "gemini-2.5-flash";
    
    await ctx.scheduler.runAfter(0, internal.chat.triggerStreaming, {
      streamId,
      prompt: args.prompt,
      model: selectedModel,
      chatId,
    });

    return { chatId, streamId };
  },
});

export const triggerStreaming = internalAction({
  args: {
    streamId: v.string(),
    prompt: v.string(),
    model: v.string(),
    chatId: v.optional(v.id("chats")),
  },  handler: async (ctx, args) => {
    // Get conversation history if we have a chat ID
    let conversationHistory: {role: string, content: string}[] = [];
    if (args.chatId) {
      try {
        const chatMessages = await ctx.runQuery(internal.chat.getChatMessagesForStreaming, {
          chatId: args.chatId
        });
        
        if (chatMessages && chatMessages.length > 0) {
          conversationHistory = chatMessages.map((msg: {role: string, content: string}) => ({
            role: msg.role,
            content: msg.content
          }));
        }
      } catch (error) {
        console.error("Error retrieving conversation history:", error);
      }
    }
    
    const convexUrl = process.env.CONVEX_SITE_URL;
    if (!convexUrl) {
      console.error("CONVEX_SITE_URL not set");
      return;
    }
    
    const requestUrl = `${convexUrl}/chat-stream`;    const requestBody = {
      streamId: args.streamId,
      prompt: args.prompt,
      model: args.model,
      conversationHistory, // Include conversation history
      chatId: args.chatId, // Include chatId so HTTP endpoint can save response directly
    };
    
    try {
      const response = await fetch(requestUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(requestBody),
      });      if (!response.ok) {
        const errorText = await response.text();
        console.error("HTTP request failed:", response.status, response.statusText, errorText);
      } else {
        console.log("Streaming triggered successfully");
      }
    } catch (error) {
      console.error("Failed to trigger streaming:", error);
    }
  },
});

export const getChatBody = query({
  args: { streamId: v.string() },
  handler: async (ctx, args) => {
    return await persistentTextStreaming.getStreamBody(ctx, args.streamId as any);
  },
});

export const getChat = query({
  args: { chatId: v.id("chats") },
  handler: async (ctx, args) => {
    return await ctx.db.get(args.chatId);
  },
});

export const getUserChats = query({
  args: {},
  handler: async (ctx) => {
    // Return all chats since we don't have auth anymore
    return await ctx.db
      .query("chats")
      .order("desc")
      .take(50);
  },
});

// Create a chat without authentication requirements
export const createUserChat = mutation({
  args: {
    prompt: v.string(),
    model: v.optional(v.string()),
  },  handler: async (ctx, args) => {
    const streamId = await persistentTextStreaming.createStream(ctx);

    // Create thread first
    const threadId = await ctx.db.insert("threads", {
      title: args.prompt.slice(0, 50) + (args.prompt.length > 50 ? "..." : ""),
      createdAt: Date.now(),
      updatedAt: Date.now(),
    });

    const chatId = await ctx.db.insert("chats", {
      title: args.prompt.slice(0, 50) + (args.prompt.length > 50 ? "..." : ""),
      prompt: args.prompt,
      streamId,
      threadId: threadId,
      createdAt: Date.now(),
      updatedAt: Date.now(),
    });    // Save user message to database with proper sequence
    await ctx.runMutation(internal.chat.saveMessageWithSequenceInternal, {
      threadId: threadId,
      content: args.prompt,
      role: "user" as const,
    });// Trigger streaming via internal action
    const selectedModel = args.model || "gemini-2.5-flash";
    
    await ctx.scheduler.runAfter(0, internal.chat.triggerStreaming, {
      streamId,
      prompt: args.prompt,
      model: selectedModel,
      chatId,
    });

    return { chatId, streamId };
  },
});

// Delete a chat
export const deleteChat = mutation({
  args: { chatId: v.id("chats") },
  handler: async (ctx, args) => {
    // Since we don't have auth, just delete the chat
    const chat = await ctx.db.get(args.chatId);
    if (!chat) {
      throw new Error("Chat not found");
    }
    
    await ctx.db.delete(args.chatId);
  },
});

// Update chat title
export const updateChatTitle = mutation({
  args: { 
    chatId: v.id("chats"), 
    title: v.string(),
  },
  handler: async (ctx, args) => {
    const chat = await ctx.db.get(args.chatId);
    if (!chat) {
      throw new Error("Chat not found");
    }
    
    await ctx.db.patch(args.chatId, { 
      title: args.title, 
      updatedAt: Date.now() 
    });
  },
});

// Get chat messages for display (for non-agent chats)
export const getChatMessages = query({
  args: { chatId: v.id("chats") },
  handler: async (ctx, args) => {
    const chat = await ctx.db.get(args.chatId);
    
    if (!chat) {
      return null;
    }
      // For streaming chats, we'll get the stream body
    if (chat.streamId) {
      const streamBody = await persistentTextStreaming.getStreamBody(ctx, chat.streamId as any);
      return {
        chat,
        streamBody,
        messages: [
          {
            id: "user-prompt",
            text: chat.prompt || "",
            isUser: true,
            timestamp: chat.createdAt,
          },
          ...(streamBody ? [{
            id: `stream-${chat._id}`,
            text: typeof streamBody === "string" ? streamBody : (streamBody as any)?.text || (streamBody as any)?.body || "",
            isUser: false,
            timestamp: chat.createdAt + 1000,
            isStreaming: false,
          }] : [])
        ]
      };
    }
    
    return { chat, messages: [] };
  },
});

export const agentChat = action({
  args: {
    prompt: v.string(),
    model: v.optional(v.string()),
    threadId: v.optional(v.string()),
  },  handler: async (ctx, args) => {
    
    try {
      let threadId = args.threadId;
      let thread;

      if (threadId) {
        // Continue existing thread
        const result = await utilityAgent.continueThread(ctx, { threadId });
        thread = result.thread;
      } else {
        // Create new thread
        const result = await utilityAgent.createThread(ctx, {});
        threadId = result.threadId;
        thread = result.thread;
      }

      // Generate response using the agent
      const result = await thread.generateText({ prompt: args.prompt });
      
      return {
        success: true,
        response: result.text,
        threadId,
        toolsUsed: result.toolCalls?.map(call => call.toolName) || [],
        usage: result.usage,
      };
      
    } catch (error) {
      console.error("Agent action failed:", error);
      return {
        success: false,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  },
});

// Get thread messages for display in the UI - using agent framework
export const getThreadMessages = action({
  args: { threadId: v.string() },
  handler: async (ctx, args) => {
    try {
      const messages = await utilityAgent.listMessages(ctx, {
        threadId: args.threadId,
        paginationOpts: { cursor: null, numItems: 50 }
      });
      
      return messages.page;
    } catch (error) {
      console.error("Error fetching thread messages:", error);
      return [];
    }
  },
});

// List all threads for a user - using component queries
export const getUserThreads = query({
  args: {},
  handler: async (ctx, args) => {
    // Since we don't have users anymore, return empty array
    // You might want to implement a different logic here based on your needs
    return [];
  },
});

// Continue an existing chat with a new message
export const continueUserChat = mutation({
  args: {
    chatId: v.id("chats"),
    prompt: v.string(),
    model: v.optional(v.string()),
  },  handler: async (ctx, args) => {
    // Verify chat exists
    const existingChat = await ctx.db.get(args.chatId);
    if (!existingChat) {
      throw new Error("Chat not found");
    }

    // Create a new stream for the continuation
    const streamId = await persistentTextStreaming.createStream(ctx);

    // Update the chat with new stream and latest prompt
    await ctx.db.patch(args.chatId, {
      streamId,
      prompt: args.prompt, // Update to latest prompt
      updatedAt: Date.now(),
    });    // Save user message to database with proper sequence
    if (existingChat.threadId) {
      await ctx.runMutation(internal.chat.saveMessageWithSequenceInternal, {
        threadId: existingChat.threadId,
        content: args.prompt,
        role: "user" as const,
      });
    } else {
      console.error("No threadId found for chat:", args.chatId);
    }

    // Trigger streaming via internal action
    const selectedModel = args.model || "gemini-2.5-flash";
    
    await ctx.scheduler.runAfter(0, internal.chat.triggerStreaming, {
      streamId,
      prompt: args.prompt,
      model: selectedModel,
      chatId: args.chatId,
    });

    return { chatId: args.chatId, streamId };
  },
});

// Save a message to the database
export const saveMessage = mutation({
  args: {
    chatId: v.id("chats"),
    content: v.string(),
    role: v.union(v.literal("user"), v.literal("assistant"), v.literal("system")),
    model: v.optional(v.string()),
  },
  handler: async (ctx, args) => {
    // Verify chat exists
    const chat = await ctx.db.get(args.chatId);
    if (!chat) {
      throw new Error("Chat not found");
    }

    // Create thread if it doesn't exist
    let threadId = chat.threadId;
    if (!threadId) {
      const newThreadId = await ctx.db.insert("threads", {
        title: chat.title,
        createdAt: Date.now(),
        updatedAt: Date.now(),
      });
      
      // Update chat with thread reference
      await ctx.db.patch(args.chatId, { threadId: newThreadId });
      threadId = newThreadId;
    }    const sequenceNumber = await getNextSequenceNumber(ctx, threadId);
    
    const messageId = await ctx.db.insert("messages", {
      threadId,
      content: args.content,
      role: args.role,
      sequenceNumber,
      metadata: args.model ? { model: args.model } : undefined,
      createdAt: Date.now(),
    });

    return messageId;
  },
});

// Save a message with proper sequence number
export const saveMessageWithSequence = mutation({
  args: {
    threadId: v.id("threads"),
    content: v.string(),
    role: v.union(v.literal("user"), v.literal("assistant"), v.literal("system")),
    model: v.optional(v.string()),
    metadata: v.optional(v.object({
      model: v.optional(v.string()),
      streamId: v.optional(v.string()),
      savedAt: v.optional(v.number()),
      lastUpdated: v.optional(v.number()),
      usage: v.optional(v.object({
        promptTokens: v.number(),
        completionTokens: v.number(),
        totalTokens: v.number(),
      })),
    })),
  },
  handler: async (ctx, args) => {
    const sequenceNumber = await getNextSequenceNumber(ctx, args.threadId);
    
    const messageId = await ctx.db.insert("messages", {
      threadId: args.threadId,
      content: args.content,
      role: args.role,
      sequenceNumber,
      metadata: args.metadata,
      createdAt: Date.now(),
    });

    return { messageId, sequenceNumber };
  },
});

// Internal version of saveMessageWithSequence
export const saveMessageWithSequenceInternal = internalMutation({
  args: {
    threadId: v.id("threads"),
    content: v.string(),
    role: v.union(v.literal("user"), v.literal("assistant"), v.literal("system")),
    model: v.optional(v.string()),
    metadata: v.optional(v.object({
      model: v.optional(v.string()),
      streamId: v.optional(v.string()),
      savedAt: v.optional(v.number()),
      lastUpdated: v.optional(v.number()),
      usage: v.optional(v.object({
        promptTokens: v.number(),
        completionTokens: v.number(),
        totalTokens: v.number(),
      })),
    })),
  },
  handler: async (ctx, args) => {
    const sequenceNumber = await getNextSequenceNumber(ctx, args.threadId);
    
    const messageId = await ctx.db.insert("messages", {
      threadId: args.threadId,
      content: args.content,
      role: args.role,
      sequenceNumber,
      metadata: args.metadata,
      createdAt: Date.now(),
    });

    return { messageId, sequenceNumber };
  },
});

// Save AI response with streaming completion tracking
export const saveAIResponseWithTracking = mutation({
  args: {
    chatId: v.id("chats"),
    content: v.string(),
    model: v.string(),
    streamId: v.string(),
  },
  handler: async (ctx, args) => {
    const chat = await ctx.db.get(args.chatId);
    if (!chat || !chat.threadId) {
      throw new Error("Chat not found or no thread ID");
    }    // Check if we already have an AI response for this stream
    const allMessages = await ctx.db
      .query("messages")
      .withIndex("by_thread", (q) => q.eq("threadId", chat.threadId as Id<"threads">))
      .filter((q) => q.eq(q.field("role"), "assistant"))
      .collect();
    
    const existingMessage = allMessages.find(msg => 
      msg.metadata?.streamId === args.streamId
    );

    if (existingMessage) {
      // Update existing message if the new content is longer (more complete)
      if (args.content.length > existingMessage.content.length) {        await ctx.db.patch(existingMessage._id, {
          content: args.content,
          metadata: { 
            model: args.model, 
            streamId: args.streamId,
            lastUpdated: Date.now()
          },
        });
        return existingMessage._id;
      }
      return existingMessage._id;
    }

    // Create new AI response message with proper sequence
    const sequenceNumber = await getNextSequenceNumber(ctx, chat.threadId as Id<"threads">);
      const messageId = await ctx.db.insert("messages", {
      threadId: chat.threadId as Id<"threads">,
      content: args.content,
      role: "assistant" as const,
      sequenceNumber,
      metadata: { 
        model: args.model, 
        streamId: args.streamId,
        savedAt: Date.now()
      },
      createdAt: Date.now(),
    });

    return messageId;
  },
});

// Get messages for a chat
export const getChatMessagesFromDB = query({
  args: { chatId: v.id("chats") },
  handler: async (ctx, args) => {
    const chat = await ctx.db.get(args.chatId);
    
    if (!chat) {
      return null;
    }

    if (!chat.threadId) {
      return { chat, messages: [] };
    }const messages = await ctx.db
      .query("messages")
      .withIndex("by_thread", (q) => q.eq("threadId", chat.threadId as Id<"threads">))
      .order("asc")
      .collect();

    return {
      chat,
      messages: messages.map(msg => ({
        id: msg._id,
        text: msg.content,
        isUser: msg.role === "user",
        timestamp: msg.createdAt,
        model: msg.metadata?.model,
      }))
    };
  },
});

// Internal mutation to save AI response (simplified)
export const saveAIResponse = internalMutation({
  args: {
    chatId: v.id("chats"),
    content: v.string(),
    model: v.string(),
  },
  handler: async (ctx, args) => {
    console.log("saveAIResponse: Called with", {
      chatId: args.chatId,
      contentLength: args.content.length,
      model: args.model
    });
    
    const chat = await ctx.db.get(args.chatId);
    if (!chat || !chat.threadId) {
      console.error("saveAIResponse: Chat not found or no thread ID for chatId:", args.chatId);
      return;
    }

    console.log("saveAIResponse: Found chat with threadId:", chat.threadId);

    // Get the next sequence number for proper ordering
    const sequenceNumber = await getNextSequenceNumber(ctx, chat.threadId);
    
    console.log("saveAIResponse: Next sequence number:", sequenceNumber);
    
    // Save AI response to database with sequence number
    const messageId = await ctx.db.insert("messages", {
      threadId: chat.threadId,
      content: args.content,
      role: "assistant" as const,
      sequenceNumber,
      metadata: { model: args.model },
      createdAt: Date.now(),
    });

    console.log("saveAIResponse: Successfully saved AI message with ID:", messageId);
    return messageId;
  },
});

// Internal query to get chat messages for streaming (includes conversation history)
export const getChatMessagesForStreaming = internalQuery({
  args: { chatId: v.id("chats") },
  handler: async (ctx, args) => {
    const chat = await ctx.db.get(args.chatId);
    
    if (!chat || !chat.threadId) {
      return [];
    }

    const messages = await ctx.db
      .query("messages")
      .withIndex("by_thread", (q) => q.eq("threadId", chat.threadId as Id<"threads">))
      .order("asc")
      .collect();

    return messages.map(msg => ({
      role: msg.role,
      content: msg.content,
      createdAt: msg.createdAt,
    }));},
});

// Helper function to get the next sequence number for a thread (with retry for race conditions)
const getNextSequenceNumber = async (ctx: any, threadId: Id<"threads">) => {
  const lastMessage = await ctx.db
    .query("messages")
    .withIndex("by_thread_sequence", (q: any) => q.eq("threadId", threadId))
    .order("desc")
    .first();
  
  return (lastMessage?.sequenceNumber || 0) + 1;
};
