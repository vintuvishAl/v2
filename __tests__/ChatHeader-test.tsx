import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';

import { ChatHeader } from '@/components/chat/ChatHeader';

describe('<ChatHeader />', () => {
  const mockOnMenuPress = jest.fn();
  const mockOnSearchPress = jest.fn();
  const mockOnNewChatPress = jest.fn();
  const mockOnToggleAgent = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  const defaultProps = {
    onMenuPress: mockOnMenuPress,
    onSearchPress: mockOnSearchPress,
    onNewChatPress: mockOnNewChatPress,
    useAgent: false,
    onToggleAgent: mockOnToggleAgent,
  };

  test('renders chat header title', () => {
    const { getByText } = render(<ChatHeader {...defaultProps} />);
    expect(getByText('V2 Chat')).toBeTruthy();
  });

  test('renders menu button', () => {
    const { getByTestId } = render(<ChatHeader {...defaultProps} />);
    expect(getByTestId('all-chats-button')).toBeTruthy();
  });

  test('renders add button', () => {
    const { getByTestId } = render(<ChatHeader {...defaultProps} />);
    expect(getByTestId('add-button')).toBeTruthy();
  });

  test('calls onMenuPress when menu button is pressed', () => {
    const { getByTestId } = render(<ChatHeader {...defaultProps} />);
    const menuButton = getByTestId('all-chats-button');
    
    fireEvent.press(menuButton);
    expect(mockOnMenuPress).toHaveBeenCalledTimes(1);
  });

  test('calls onNewChatPress when add button is pressed', () => {
    const { getByTestId } = render(<ChatHeader {...defaultProps} />);
    const addButton = getByTestId('add-button');
    
    fireEvent.press(addButton);
    expect(mockOnNewChatPress).toHaveBeenCalledTimes(1);
  });

  test('works with useAgent enabled', () => {
    const { getByText } = render(<ChatHeader {...defaultProps} useAgent={true} />);
    expect(getByText('V2 Chat')).toBeTruthy();
  });

  test('works with useAgent disabled', () => {
    const { getByText } = render(<ChatHeader {...defaultProps} useAgent={false} />);
    expect(getByText('V2 Chat')).toBeTruthy();
  });

  test('matches snapshot', () => {
    const tree = render(<ChatHeader {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
