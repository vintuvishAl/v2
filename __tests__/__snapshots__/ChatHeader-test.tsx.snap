// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatHeader /> matches snapshot 1`] = `
<View
  className="flex-row justify-between items-center px-6 py-4 bg-app-dark-background"
>
  <View
    className="flex-row items-center"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="all-chats-button"
    >
      <Text
        allowFontScaling={false}
        className="text-app-dark-icon"
        selectable={false}
        style={
          [
            {
              "color": "#1cb8ff",
              "fontSize": 30,
            },
            undefined,
            {
              "fontFamily": "ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="add-button"
    >
      <Text
        allowFontScaling={false}
        className="text-app-dark-icon"
        selectable={false}
        style={
          [
            {
              "color": "#1cb8ff",
              "fontSize": 30,
            },
            undefined,
            {
              "fontFamily": "ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
  <View
    className="flex-row items-center space-x-2"
  >
    <Text
      className="text-app-dark-text text-xl font-bold"
    >
      V2 Chat
    </Text>
  </View>
</View>
`;
